### PR TITLE
Action buttons will now only update when needed instead of every Life().

### PR DESCRIPTION
**Pull request by phil235**
- The action buttons now update their icon instantly.
- Fixes versions of pickup(),equipped() and dropped not calling the parent.
- Fixes drone not being able to remove a defib from their storage.
- You can now cycle the mime mask by clicking it in your hand.
- The action buttons for hardsuit and hooded suits now only appears when you're wearing the suit.
- Created two mob helper procs getBeltSlot() and getBackSlot().
- Created /datum/species/proc/on_species_loss() to handle stuff when our race change, currently only used by jelly and slime race to remove their exotic blood from our reagents and to remove slime people's action buttons.
- <b>Inserting items into a mob's chest no longer gives that item's action button to that mob (it was a bug).</b>
- Refactored the item's action system. Items can now hold multiple actions more easily.
- The "set internals" button of tank items now turn green when it's used as internals.
- <b>Removed research scanner from drones (since cyborgs don't have it, it's more consistent)</b>
- Removed the ignore_madkadjust mask var.
- The sechailer mask now has an adjust mask action button, so I removed the adjust verb that it was using.
- The item's action are now created on item/New() instead of trying to create it every time someone picks the item up.
- I split hud/action.dm, the datum/action stuff is now in the datum folder (/datum/action.dm), whereas the code for action buttons is kept in the hud folder under action_button.dm. Also I moved some /datum/action code that was in some files back into datum/action.dm where it belongs.

:cl: phil235
rscadd: The action button of gas tanks and jetpacks now turn green when you turn your internals on.
/:cl:

### CODE REVIEW

File: code/_onclick/hud/action.dm [hunk 1]
```diff
@@ -44,37 +50,28 @@
...
 /datum/action/proc/Process()
 	return
 
-/datum/action/proc/CheckRemoval(mob/living/user) // 1 if action is no longer valid for this mob and should be removed
-	return 0
-
 /datum/action/proc/IsAvailable()
-	return Checks()
-
-/datum/action/proc/Checks()// returns 1 if all checks pass
```
Comment (phil235) [inline code/_onclick/hud/action.dm]: I'm merging Checks() and IsAvailable(), they do the same thing.

File: code/_onclick/hud/action.dm [hunk 2]
```diff
@@ -44,37 +50,28 @@
...
 	if(!owner)
 		return 0
 	if(check_flags & AB_CHECK_RESTRAINED)
 		if(owner.restrained())
 			return 0
 	if(check_flags & AB_CHECK_STUNNED)
-		if(owner.stunned)
+		if(owner.stunned || owner.weakened)
```
Comment (phil235) [inline code/_onclick/hud/action.dm]: if you can't use the button when stunned, you sure as hell can't use it when weakened.

File: code/_onclick/hud/action.dm [hunk 3]
```diff
@@ -44,37 +50,28 @@
...
 			return 0
 	if(check_flags & AB_CHECK_LYING)
 		if(owner.lying)
 			return 0
-	if(check_flags & AB_CHECK_ALIVE)
+	if(check_flags & AB_CHECK_CONSCIOUS)
```
Comment (phil235) [inline code/_onclick/hud/action.dm]: This checks if the mob is conscious, why call it check_alive then??

File: code/_onclick/hud/action.dm [hunk 4]
```diff
@@ -90,8 +87,8 @@
 		current_button.overlays += img
 
 /obj/screen/movable/action_button
-	var/datum/action/owner
-	screen_loc = "WEST,NORTH"
+	var/datum/action/linked_action
```
Comment (phil235) [inline code/_onclick/hud/action.dm]: renamed to linked_action since "owner" is confusing, especially since /datum/action has its own owner var that references a mob.

File: code/game/objects/items.dm [hunk 5]
```diff
@@ -387,7 +388,20 @@ var/global/image/fire_overlay = image("icon" = 'icons/effects/fire.dmi', "icon_s
 // for items that can be placed in multiple slots
 // note this isn't called during the initial dressing of a player
 /obj/item/proc/equipped(mob/user, slot)
-	return
+	if(action_button_name)
```
Comment (RemieRichards) [inline code/game/objects/items.dm]: No, stop.
Stop supporting this system.

it needs to DIE.

File: code/_onclick/hud/action.dm [hunk 6]
```diff
@@ -90,8 +87,8 @@
...
+	screen_loc = null
```
Comment (phil235) [inline code/_onclick/hud/action.dm]: Need screen_loc to start null, I don't remember why...

**Comment by ChangelingRain** [comment]
Phil is based as fuck.

File: code/_onclick/hud/movable_screen_objects.dm [hunk 7]
```diff
@@ -43,7 +43,7 @@
 		var/pix_Y = text2num(screen_loc_Y[2]) - 16
 		screen_loc = "[screen_loc_X[1]]:[pix_X],[screen_loc_Y[1]]:[pix_Y]"
 
-	moved = TRUE
+	moved = screen_loc
```
Comment (phil235) [inline code/_onclick/hud/movable_screen_objects.dm]: might as well use the var to remember the screen_loc position.

File: code/game/gamemodes/handofgod/actions.dm [hunk 8]
```diff
@@ -19,3 +19,7 @@
 		return
 	god << "<span class='notice'><B>[owner]:</B> [msg]</span>"
 	owner << "You say: [msg]"
+
+/datum/action/innate/godspeak/Destroy()
+	god = null
```
Comment (phil235) [inline code/game/gamemodes/handofgod/actions.dm]: clearing the reference to the god mob.

File: code/game/gamemodes/handofgod/god.dm [hunk 9]
```diff
@@ -51,6 +51,10 @@
 	for(var/datum/mind/F in followers)
 		if(F.current)
 			F.current << "<span class='danger'>Your god is DEAD!</span>"
+	for(var/X in prophet_hats)
+		var/obj/item/clothing/head/helmet/plate/crusader/prophet/P = X
+		if(P.speak2god && P.speak2god.god == src)
+			qdel(P.speak2god)
```
Comment (phil235) [inline code/game/gamemodes/handofgod/god.dm]: When the god dies every prophet hat loses its speak action.

File: code/game/gamemodes/miniantags/abduction/abduction_gear.dm [hunk 10]
```diff
@@ -29,20 +29,19 @@
...
-
+	if(istype(loc, /mob/living/carbon/human))
+		var/mob/living/carbon/human/H = loc
+		H.update_inv_wear_suit()
+	if(action && action.button)
+		action.button.UpdateIcon()
+
+/obj/item/clothing/suit/armor/abductor/vest/item_action_slot_check(slot, mob/user)
+	if(slot == slot_wear_suit) //we only give the mob the ability to activate the vest if he's actually wearing it.
+		return 1
```
Comment (phil235) [inline code/game/gamemodes/miniantags/abduction/abduction_gear.dm]: This proc is used so the action button of the vest only shows up when worn and not just in your hand.

File: code/game/gamemodes/miniantags/morph/morph.dm [hunk 11]
```diff
@@ -169,9 +169,6 @@
 			return
 	target.attack_animal(src)
 
-/mob/living/simple_animal/hostile/morph/update_action_buttons() //So all eaten objects are not counted every life
-	return
-
```
Comment (phil235) [inline code/game/gamemodes/miniantags/morph/morph.dm]: no longer needed, we don't check inside the mob for items' action button.

File: code/game/objects/items/weapons/defib.dm [hunk 12]
```diff
@@ -70,35 +70,40 @@
...
+			else
+				user << "<span class='warning'>Strap the defibrillator's belt on first!</span>"
 		return
 	..()
 
 /obj/item/weapon/defibrillator/MouseDrop(obj/over_object)
-	if(ishuman(src.loc))
-		var/mob/living/carbon/human/H = src.loc
+	if(ismob(src.loc))
+		var/mob/M = src.loc
```
Comment (phil235) [inline code/game/objects/items/weapons/defib.dm]: so drones can remove the defib from their storage.

File: code/game/objects/items/weapons/defib.dm [hunk 13]
```diff
@@ -172,16 +177,22 @@
...
 /obj/item/weapon/defibrillator/equipped(mob/user, slot)
-	if(slot != slot_back)
+	..()
+	if((slot_flags == SLOT_BACK && slot != slot_back) || (slot_flags == SLOT_BELT && slot != slot_belt))
 		remove_paddles(user)
 		update_icon()
 
+/obj/item/weapon/defibrillator/item_action_slot_check(slot, mob/user)
+	if(slot == user.getBackSlot())
+		return 1
```
Comment (phil235) [inline code/game/objects/items/weapons/defib.dm]: Using getBackSlot means drones can actually deploy the paddles.

File: code/game/objects/items/weapons/implants/implant_freedom.dm [hunk 14]
```diff
@@ -8,14 +8,13 @@
...
-		return 0
-	if(uses != -1)
-		uses--
+	uses--
 	imp_in << "You feel a faint click."
 	if(iscarbon(imp_in))
 		var/mob/living/carbon/C_imp_in = imp_in
 		C_imp_in.uncuff()
+	if(!uses)
+		qdel(src)
```
Comment (phil235) [inline code/game/objects/items/weapons/implants/implant_freedom.dm]: When the implant is out of use we delete it which removes the action button from the mob.

File: code/game/objects/items/weapons/storage/storage.dm [hunk 15]
```diff
@@ -374,10 +374,6 @@
 	handle_item_insertion(W, 0 , user)
 	return 1
 
-
-/obj/item/weapon/storage/dropped(mob/user)
-	return
-
```
Comment (phil235) [inline code/game/objects/items/weapons/storage/storage.dm]: Why was this here? Nobody knows.

File: code/game/objects/items/weapons/tanks/tanks.dm [hunk 16]
```diff
@@ -1,50 +1,41 @@
...
 	hitsound = 'sound/weapons/smash.ogg'
-
 	pressure_resistance = ONE_ATMOSPHERE * 5
-
 	force = 5
 	throwforce = 10
 	throw_speed = 1
 	throw_range = 4
-
+	action_button_name = "Set Internals"
```
Comment (phil235) [inline code/game/objects/items/weapons/tanks/tanks.dm]: the tank's action to set internals now use item/var/action instead of creating a new variable.

File: code/game/objects/items/weapons/tanks/tanks.dm [hunk 17]
```diff
@@ -1,50 +1,41 @@
...
+/obj/item/weapon/tank/ui_action_click()
+	var/mob/living/carbon/human/H = action.owner
+	if(!istype(H))
+		return
+
+	if(!H.wear_mask)
+		H << "<span class='warning'>You need a mask!</span>"
+		return
+	if(H.wear_mask.mask_adjusted)
+		H.wear_mask.adjustmask(H)
```
Comment (phil235) [inline code/game/objects/items/weapons/tanks/tanks.dm]: when using the internals icon (the one above the health hud) the code tries to adjust the mask back up before trying to turn on internals, but the tank's action button wasn't doing it. This fixes that.

File: code/game/objects/items/weapons/tanks/watertank.dm [hunk 18]
```diff
@@ -75,22 +80,21 @@
 	..()
 
 /obj/item/weapon/watertank/MouseDrop(obj/over_object)
-	var/mob/H = src.loc
```
Comment (phil235) [inline code/game/objects/items/weapons/tanks/watertank.dm]: Let's make a mob, and...let's call it H!

File: code/game/objects/items.dm (hunk 5 above)
Comment (optimumtact) [inline code/game/objects/items.dm]: if checks?

File: code/modules/mob/living/life.dm [hunk 19]
```diff
@@ -49,10 +49,6 @@
 		handle_disabilities() // eye, ear, brain damages
 		handle_status_effects() //all special effects, stunned, weakened, jitteryness, hallucination, sleeping, etc
 
-	handle_actions()
-
-	handle_regular_hud_updates()
-
```
Comment (phil235) [inline code/modules/mob/living/life.dm]: The whole reason for this PR. Removing two lines.

File: code/modules/mob/living/silicon/robot/life.dm [hunk 20]
```diff
@@ -30,7 +30,7 @@
 		update_headlamp()
 	diag_hud_set_borgcell()
 
-/mob/living/silicon/robot/handle_regular_hud_updates()
+/mob/living/silicon/robot/proc/handle_robot_hud_updates()
```
Comment (phil235) [inline code/modules/mob/living/silicon/robot/life.dm]: this proc is the next on my list of shit to remove from Life()

File: code/modules/mob/living/simple_animal/slime/death.dm [hunk 21]
```diff
@@ -9,8 +9,6 @@
 			M.regenerate_icons()
 			is_adult = 0
 			maxHealth = 150
-			var/datum/action/innate/slime/evolve/E = new
-			E.Grant(src)
```
Comment (phil235) [inline code/modules/mob/living/simple_animal/slime/death.dm]: crap, that's a mistake, will fix.

File: code/game/objects/items.dm (hunk 5 above)
Comment (RemieRichards) [inline code/game/objects/items.dm]: @optimumtact no...
The whole "By default items have one action datum reference, a var to control that one reference's type and a string to decide its name" system.

Its shit.
Action Datums? 100%, Amazing, Greatness.
Action Strings? get_out%, die, you suck

Comment (phil235) [inline code/game/objects/items.dm]: Alright, I'll make a proper system.

**Comment by Core0verload** [comment]
>Inserting items into a mob's chest no longer gives that item's action button to that mob (it was a bug).

NO IT WASNT. It was totally a feature!

**Comment by duncathan** [comment]
I disagree with removing that bug
It's been around long enough to call it a feature

**Comment by Iamgoofball** [comment]
do I need to start implanting maglites in my chest and then shadowling hunting for you guys to stop complaining

**Comment by duncathan** [comment]
That's fucking awesome tho
I want to enable that it's cool

**Comment by duncathan** [comment]
Anyway if you want to remove it that's fine but I disagree with it being in what's otherwise objectively a good pr

File: code/_onclick/hud/screen_objects.dm [hunk 22]
```diff
@@ -106,54 +106,59 @@
...
-						C << "<span class='notice'>You are now running on internals from the [C.back] on your back.</span>"
-						C.internal = C.back
-
-					if(C.internal)
-						icon_state = "internal1"
-					else
-						C << "<span class='warning'>You don't have an oxygen tank!</span>"
+				C << "<span class='warning'>You don't have an oxygen tank!</span>"
+				return
+	C.update_action_buttons_icon()
```
Comment (phil235) [inline code/_onclick/hud/screen_objects.dm]: this is the only real change in this file, we update the action buttons icon (since the internal tank button turns green when used as internals now).

File: code/datums/action.dm [hunk 23]
```diff
@@ -0,0 +1,357 @@
...
+	name = "Pull The Starting Cord"
+
+/datum/action/item_action/toggle_gunlight
+	name = "Toggle Gunlight"
+
+/datum/action/item_action/toggle_mode
+	name = "Toggle Mode"
+
+/datum/action/item_action/toggle_barrier_spread
+	name = "Toggle Barrier Spread"
```
Comment (phil235) [inline code/datums/action.dm]: Most of these are just used to set the appropriate name for the action button.

File: code/datums/action.dm [hunk 24]
```diff
@@ -0,0 +1,357 @@
...
+
+/datum/action/item_action/set_internals
+	name = "Set Internals"
+
+/datum/action/item_action/set_internals/UpdateButtonIcon()
+	if(..()) //button available
+		if(iscarbon(owner))
+			var/mob/living/carbon/C = owner
+			if(target == C.internal)
+				button.icon_state = "bg_internals_on"
```
Comment (phil235) [inline code/datums/action.dm]: background turns green when tank is used as internals.

File: code/datums/action.dm [hunk 25]
```diff
@@ -0,0 +1,357 @@
...
+
+/datum/action/item_action/toggle_research_scanner/Remove(mob/living/L)
+	if(owner)
+		owner.research_scanner = 0
+	..()
+
+/datum/action/item_action/toggle_research_scanner/ApplyIcon(obj/screen/movable/action_button/current_button)
+	if(button_icon && button_icon_state)
+		var/image/img = image(button_icon, current_button, "scan_mode")
+		current_button.overlays += img
```
Comment (phil235) [inline code/datums/action.dm]: I changed the research scanner action to be an item action instead of a weird datum/action/innate.

File: code/game/gamemodes/blob/powers.dm [hunk 26]
```diff
@@ -248,14 +248,6 @@
 /datum/action/innate/blob
 	background_icon_state = "bg_alien"
 
-/datum/action/innate/blob/CheckRemoval()
-	if(ticker.mode.name != "blob" || !ishuman(owner))
-		return 1
-	var/datum/game_mode/blob/B = ticker.mode
-	if(!owner.mind || !(owner.mind in B.infected_crew))
-		return 1
-	return 0
```
Comment (phil235) [inline code/game/gamemodes/blob/powers.dm]: CheckRemoval() proc is completely removed, we no longer need to check this every tick.

File: code/game/objects/items.dm [hunk 27]
```diff
@@ -33,11 +33,8 @@ var/global/image/fire_overlay = image("icon" = 'icons/effects/fire.dmi', "icon_s
...
 	var/min_cold_protection_temperature //Set this variable to determine down to which temperature (IN KELVIN) the item protects against cold damage. 0 is NOT an acceptable number due to if(varname) tests!! Keep at null to disable protection. Only protects areas set by cold_protection flags
 
-	//If this is set, The item will make an action button on the player's HUD when picked up.
-	var/action_button_name //It is also the text which gets displayed on the action button. If not set it defaults to 'Use [name]'. If it's not set, there'll be no button.
-	var/action_button_is_hands_free = 0 //If 1, bypass the restrained, lying, and stunned checks action buttons normally test for
-	var/action_button_internal = 0 //If 1, bypass the inside check action buttons test for
-	var/datum/action/item_action/action = null
+	var/list/actions = list() //list of /datum/action's that this item has.
+	var/list/actions_types = list() //list of paths of action datums to give to the item on New().
```
Comment (phil235) [inline code/game/objects/items.dm]: less variables for /obj/item.

File: code/datums/action.dm [hunk 28]
```diff
@@ -0,0 +1,357 @@
...
+		L.client.screen += button
+	L.update_action_buttons()
+
+/datum/action/proc/Remove(mob/living/L)
+	if(L.client)
+		L.client.screen -= button
+	button.moved = FALSE //so the button appears in its normal position when given to another owner.
+	L.actions -= src
+	L.update_action_buttons()
+	owner = null
```
Comment (phil235) [inline code/datums/action.dm]: We no longer delete and recreate the action's button every time it's removed and granted again.

File: code/modules/clothing/clothing.dm [hunk 29]
```diff
@@ -116,7 +116,6 @@ BLIND     // can't see anything
 	strip_delay = 40
 	put_on_delay = 40
 	var/mask_adjusted = 0
-	var/ignore_maskadjust = 1
```
Comment (phil235) [inline code/modules/clothing/clothing.dm]: this var was originally added because one single mask needed it...

File: code/modules/clothing/glasses/engine_goggles.dm [hunk 30]
```diff
@@ -30,11 +27,15 @@
 		loc << "<span class='notice'>You toggle the goggles' scanning mode to \[Meson].</span>"
 		invis_update()
 
-	if(istype(loc,/mob/living/carbon))
-		var/mob/living/carbon/C = loc
-		C.update_sight()
+	if(ishuman(user))
+		var/mob/living/carbon/human/H = user
+		if(H.glasses == src)
+			H.update_sight()
```
Comment (phil235) [inline code/modules/clothing/glasses/engine_goggles.dm]: the check is ishuman instead of iscarbon, since only humans can wear glasses.

File: code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm [hunk 31]
```diff
@@ -5,34 +5,6 @@ These are general powers. Specific powers are stored under the appropriate alien
...
-/datum/action/spell_action/alien/CheckRemoval()
-	if(!iscarbon(owner))
-		return 1
-
-	var/mob/living/carbon/C = owner
-	if(target.loc && !(target.loc in C.internal_organs))
-		return 1
-
-	return 0
-
```
Comment (phil235) [inline code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm]: moved to datum/action.dm

File: code/modules/mob/living/simple_animal/friendly/drone/_drone.dm [hunk 32]
```diff
@@ -80,7 +80,6 @@
 		equip_to_slot_or_del(I, slot_head)
 
 	access_card.flags |= NODROP
-	scanner.Grant(src)
```
Comment (phil235) [inline code/modules/mob/living/simple_animal/friendly/drone/_drone.dm]: drones lose their scanner. (consistency with robots not having it, and also makes things easier codewise)

File: code/modules/mob/mob_defines.dm [hunk 33]
```diff
@@ -84,7 +84,6 @@
 	var/datum/hud/hud_used = null
 
 	var/research_scanner = 0 //For research scanner equipped mobs. Enable to show research data when examining.
-	var/datum/action/innate/scan_mode/scanner = new
```
Comment (phil235) [inline code/modules/mob/mob_defines.dm]: Ughhhhh....

File: code/game/objects/items.dm [hunk 34]
```diff
@@ -131,10 +128,17 @@ var/global/image/fire_overlay = image("icon" = 'icons/effects/fire.dmi', "icon_s
 /obj/item/device
 	icon = 'icons/obj/device.dmi'
 
+/obj/item/New()
+	..()
+	for(var/path in actions_types)
+		new path(src)
```
Comment (phil235) [inline code/game/objects/items.dm]: We create the item's actions in New() by going through the actions_types list.

**Comment by phil235** [comment]
Ready for review

File: code/game/objects/items.dm (hunk 27 above)
Comment (AnturK) [inline code/game/objects/items.dm]: :+1: Never got to doing this.

**Comment by Core0verload** [comment]
@phil235, how about adding implanted item action button back?

File: code/game/objects/items.dm (hunk 27 above)
Comment (RemieRichards) [inline code/game/objects/items.dm]: @phil235 you could actually just use the same list.
(removing, newing, adding back to the list)

File: code/datums/action.dm [hunk 35]
```diff
@@ -0,0 +1,357 @@
...
+	var/obj/item/I = target
+	I.actions -= src
+	return ..()
+
+/datum/action/item_action/Trigger()
+	if(!..())
+		return 0
+	if(target)
+		var/obj/item/I = target
+		I.ui_action_click(owner, src.type)
```
Comment (RemieRichards) [inline code/datums/action.dm]: next goal would be to destroy ui_action_click(), but that can be done at another time.

File: code/datums/action.dm [hunk 36]
```diff
@@ -0,0 +1,357 @@
...
+	if(button_icon && button_icon_state)
+		var/image/img = image(button_icon, current_button, "scan_mode")
+		current_button.overlays += img
+
+/datum/action/item_action/organ_action
+	check_flags = AB_CHECK_CONSCIOUS
+
+/datum/action/item_action/organ_action/New(Target)
+	..()
+	name = "Toggle [target.name]"
```
Comment (RemieRichards) [inline code/datums/action.dm]: All organ actions are now "Toggle [organname]"? Why?

**Comment by RemieRichards** [comment]
@phil235 Besides what I've commented, this is good.

File: code/datums/action.dm (hunk 36 above)
Comment (phil235) [inline code/datums/action.dm]: Because the current button names used by organ actions are all pretty much "Toggle [theorganname]" , check the old "organ_action_name = "stuff" "

Comment (RemieRichards) [inline code/datums/action.dm]: Yes except you CANT give it a new name now.

it should be ```if(!name) name = "Toggle [target.name]"```

**Comment by phil235** [comment]
>how about adding implanted item action button back?

Most actions shouldn't be accessible when inside your chest. Especially flashlights which are used to powergame against shadowlings (plus emitting light through your chest is not realistic at all). I haven't seen any item action that was worth keeping when inserted in you.

**Comment by Core0verload** [comment]
>Most actions shouldn't be accessible when inside your chest. Especially flashlights which are used to powergame against shadowlings (plus emitting light through your chest is not realistic at all)

Do you really think a surgeon of SS13 can't implant the flashlight while leaving a lightbulb outside and "on" switch accessible? Do you really thing the concept of "realism" justifies removing fun features?

I'll bring action button for implanted items back myself, but with new system I don't even know how.

**Comment by optimumtact** [comment]
people can still do that, it just has to be explicitly coded instead of just being a bug that granted people action buttons to stuff that was in their contents.
